## CaseStudy & Problems

In [1]:
import numpy as np

In [2]:
!head cric_data.tsv

'head' is not recognized as an internal or external command,
operable program or batch file.


1. Find mean, median, IQR for Sachin , Rahul and India
2. Find the histogram of Scahin's scores with 10 bins
3. Find mean of Schin's scores grouped by 25 matches
4. Find mean for Sachin's scores where he has score a century
5. Find mean of Sachin's scores when Rahul has scored less than 10
6. Find mean for Sachin's scores based on which quartile India's score falls in
7. For every match find out who has has scored more - Sachin or Rahul
8.  How many more runs does Sachin score on average after having scored x runs
9. How many matches did Sachin take to score first 1000 runs, next 1000 runs,....

In [3]:
# cric_data = np.loadtxt('cric_data.tsv')

In [4]:
cric_data = np.loadtxt('cric_data.tsv', skiprows=1)

In [5]:
cric_data.shape

(225, 4)

In [6]:
cric_data=cric_data[:, [1, 2, 3]]

In [7]:
cric_data.shape

(225, 3)

In [8]:
sachin = cric_data[:, 0] 

In [9]:
rahul = cric_data[:, 1]

In [10]:
india = cric_data[:, 2]

**1. Find mean, median, IQR for Sachin , Rahul and India**

In [11]:
def stats(cols):
    print("mean : ",np.mean(cols))
    print("median : ", np.median(cols))
    print("IQR : ",np.percentile(cols, 75)-np.percentile(cols, 25))

In [12]:
stats(sachin)

mean :  39.87555555555556
median :  27.0
IQR :  57.0


In [13]:
stats(rahul)

mean :  32.062222222222225
median :  22.0
IQR :  46.0


In [14]:
stats(india)

mean :  220.79555555555555
median :  216.0
IQR :  98.0


In [15]:
np.mean(cric_data, axis = 0)

array([ 39.87555556,  32.06222222, 220.79555556])

In [16]:
np.median(cric_data, axis = 0)

array([ 27.,  22., 216.])

In [17]:
np.percentile(cric_data, 75, axis = 0) - np.percentile(cric_data, 25, axis = 0)

array([57., 46., 98.])

**2.Find the histogram of Scahin's scores with 10 bins**

In [18]:
np.histogram(sachin, bins=10)

(array([99, 36, 28, 16, 11, 17,  8,  8,  1,  1], dtype=int64),
 array([  0. ,  18.6,  37.2,  55.8,  74.4,  93. , 111.6, 130.2, 148.8,
        167.4, 186. ]))

**3.Find mean of Schin's scores grouped by 25 matches**

In [19]:
sachin.shape

(225,)

In [20]:
sachin.reshape(9, 25).shape

(9, 25)

In [21]:
sachin_by_25s = sachin.reshape(9, 25)

In [22]:
np.mean(sachin_by_25s, axis=1)

array([33.96, 49.4 , 38.48, 40.16, 39.36, 38.2 , 44.6 , 39.52, 35.2 ])

**4.Find mean for Sachin's scores where he has score a century**

In [23]:
sachin >= 100

array([ True, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
       False,  True, False, False, False, False, False, False,  True,
       False, False,  True, False, False, False, False, False, False,
       False, False,  True,  True, False, False, False, False, False,
       False,  True, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [24]:
sachin[sachin >= 100]

array([100., 104., 138., 141., 186., 146., 141., 123., 120., 122., 100.,
       152., 105., 122., 100., 117., 141., 139., 114., 127., 110., 146.,
       101., 140., 113., 102.])

In [25]:
np.mean(sachin[sachin >= 100])

125.0

**5.Find mean of Sachin's scores when Rahul has scored less than 10**

In [26]:
rahul <= 10

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True,  True, False,  True,  True,
       False, False,  True, False, False,  True, False, False,  True,
        True, False,  True, False,  True, False, False, False, False,
       False, False, False,  True,  True,  True,  True, False, False,
       False, False, False, False,  True, False, False, False, False,
        True, False, False, False, False, False, False, False,  True,
        True, False,  True,  True, False,  True, False, False,  True,
       False, False, False, False, False, False, False,  True,  True,
       False, False,  True, False,  True, False, False, False, False,
        True,  True,  True, False,  True, False, False, False, False,
       False,  True, False, False, False, False, False, False,  True,
        True, False,  True,  True, False,  True, False, False,  True,
        True, False, False, False, False,  True, False, False, False,
        True, False,

In [27]:
sachin[rahul <= 10]

array([ 37.,  14.,   4.,   0.,  62.,  38.,  65.,   0.,  48.,  62.,  27.,
        27.,  51.,  18.,  65.,  28.,   2.,  54.,   4.,  14., 100.,  57.,
         0.,   5.,   0.,  30.,  83.,  93.,   0., 152.,   0.,   1.,  53.,
         0.,  81.,  78., 122.,   9.,  13.,  36.,  29.,  12.,  34., 100.,
        44.,  82.,   0.,   6.,  49.,  64.,  43.,  72.,   3.,  44.,  47.,
        17.,  35.,  88.,   0.,  33., 110., 146.,   7.,  11.,  94.,  55.,
         0.,  28.,   2.,  27.,  40.])

In [28]:
np.mean(sachin[rahul <= 10])

40.2112676056338

**6.Find mean for Sachin's scores based on which quartile India's score falls in**

In [29]:
qrs = np.percentile(india, [25,50,75,100])

If India <=175: Sachin's average is ...

If India <=216: Sachin's average is ...

If India <=273: Sachin's average is ...

If India <=499: Sachin's average is ...

In [30]:
# india < qrs

In [31]:
india.shape

(225,)

In [32]:
qrs.shape

(4,)

In [33]:
qrs = qrs.reshape(4,1)

In [34]:
india < qrs

array([[False, False, False, False, False, False, False, False, False,
        False,  True, False, False, False,  True, False, False,  True,
        False, False,  True, False, False, False, False,  True,  True,
         True,  True,  True, False, False, False, False, False, False,
         True, False,  True, False, False, False, False, False, False,
        False, False, False, False,  True,  True, False, False, False,
         True, False, False, False, False,  True, False, False, False,
        False, False,  True, False, False, False,  True, False, False,
        False, False, False, False, False, False,  True, False,  True,
         True, False, False, False, False, False,  True, False, False,
        False, False,  True, False, False, False, False, False, False,
        False,  True,  True,  True, False, False, False, False, False,
        False, False,  True,  True, False, False, False, False, False,
        False, False, False, False, False,  True, False, False,  True,
      

In [35]:
indices = india < qrs

In [36]:
indices.shape

(4, 225)

In [37]:
sachin[indices[0, :]]

array([18., 14.,  0., 62., 46., 65.,  0., 39., 48.,  3., 11., 65., 27.,
       28.,  3.,  4., 15., 40.,  5.,  8., 89.,  0.,  0.,  1.,  0.,  0.,
       81., 13.,  2., 36., 12., 19.,  0.,  6., 35.,  0., 44.,  3., 47.,
       17., 35., 33.,  7.,  9.,  2., 11., 17.,  1., 10.,  0., 23.,  1.,
        2., 25.,  0.])

In [38]:
for i in range(4):
    print(i, np.mean(sachin[indices[i]]))

0 19.672727272727272
1 28.18018018018018
2 31.688622754491018
3 39.799107142857146


**7.For every match find out who has has scored more - Sachin or Rahul**

In [39]:
snr = cric_data[:, 0:2]

In [40]:
np.max([1,3,2,5,1])

5

In [41]:
np.argmax([1,3,2,5,1])   #largest no position find

3

In [42]:
np.argmax(snr, axis=1)

array([0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       0, 0, 1, 0, 0], dtype=int64)

In [43]:
is_rahul_higher = np.argmax(snr, axis=1)

In [44]:
np.sum(is_rahul_higher) / 225

0.4711111111111111

In [45]:
np.where(is_rahul_higher == 0, 'Rahul', 'Sachin')

array(['Rahul', 'Sachin', 'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Sachin',
       'Rahul', 'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Sachin', 'Rahul',
       'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Rahul',
       'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Rahul',
       'Rahul', 'Rahul', 'Rahul', 'Rahul', 'Sachin', 'Sachin', 'Rahul',
       'Sachin', 'Sachin', 'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Rahul',
       'Rahul', 'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Sachin',
       'Rahul', 'Sachin', 'Sachin', 'Sachin', 'Sachin', 'Rahul', 'Sachin',
       'Rahul', 'Rahul', 'Sachin', 'Sachin', 'Sachin', 'Rahul', 'Sachin',
       'Rahul', 'Rahul', 'Rahul', 'Rahul', 'Sachin', 'Rahul', 'Sachin',
       'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Sachin',
       'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Sachin', 'Sachin',
       'Sachin', 'Sachin', 'Rahul', 'Rahul', 'Rahul', 'Sachin', 'Rahul',
       'Rahul', 'Rahul', 'Rahul', 'Rahul', 'Rahul', '

**8.How many more runs does Sachin score on average after having scored x runs**

In [46]:
 x_arr = np.arange(0, 101, 5)

In [47]:
# sachin >= x_arr

In [48]:
x_arr = x_arr.reshape(x_arr.shape[0], 1)

In [49]:
indices = sachin >= x_arr

In [50]:
indices.shape

(21, 225)

In [51]:
sachin[indices[1, :]]

array([100.,  11.,   8.,  71., 104.,  18.,   8.,  86.,  12.,  85.,  18.,
         7.,  37.,  14.,  21.,  62., 138.,  38.,  46.,  65.,  39.,  48.,
       141.,  62.,  12.,  41.,  11., 186.,  11.,  27.,  27.,  51.,  18.,
        32., 146.,   5.,  45., 141.,  12.,  65.,  27.,   7.,  16.,  28.,
         6., 123., 120.,   7.,  81.,  54., 122.,  14., 100.,  15.,  57.,
        99.,  37.,  38.,  32.,  21.,  32.,  40.,   5.,   8.,   5.,  50.,
        30.,  37.,  89.,  98.,  83.,  93.,  52., 152.,   8.,  93.,  45.,
        26.,  16.,  47.,  89.,  53.,  16.,  81.,  14.,  78.,   6., 105.,
       122.,   9.,   8.,  28.,  35.,  69.,  13.,  97.,  93.,  36.,  39.,
        29.,  12.,  19.,  34., 100.,  44.,  82.,  79.,   6.,   9.,   8.,
        23.,  93.,  35.,  63.,  74.,   8., 117.,  39.,  49.,  64.,  43.,
        72.,   5.,  17.,  65.,  20., 141.,  28.,  44.,  27.,  60.,  68.,
       139.,  31.,  44.,  47.,   6.,  17.,  35.,  88., 114.,   7., 127.,
        45.,  33., 110., 146.,   7.,  25.,   9.,  1

In [52]:
for i in range(x_arr.shape[0]):
    print(x_arr[i], np.mean(sachin[indices[i, :]])- x_arr[i,0])

[0] 39.87555555555556
[5] 45.61363636363637
[10] 47.48026315789474
[15] 47.45255474452555
[20] 46.824
[25] 44.10084033613445
[30] 45.13461538461539
[35] 43.24742268041237
[40] 44.05882352941177
[45] 43.41558441558442
[50] 43.98529411764706
[55] 42.317460317460316
[60] 38.67213114754098
[65] 37.654545454545456
[70] 37.08163265306122
[75] 34.347826086956516
[80] 30.75
[85] 28.650000000000006
[90] 27.400000000000006
[95] 26.433333333333337
[100] 25.0


**9.How many matches did Sachin take to score first 1000 runs, next 1000 runs,....**

In [53]:
sachin

array([100.,  11.,   8.,  71., 104.,  18.,   8.,  86.,  12.,  85.,  18.,
         4.,   7.,  37.,  14.,   0.,   4.,   0.,  21.,   1.,  62.,   0.,
       138.,  38.,   2.,  46.,  65.,   0.,  39.,  48., 141.,  62.,  12.,
         1.,  41.,  11.,   3., 186.,  11.,  27.,  27.,  51.,  18.,  32.,
       146.,   5.,  45., 141.,  12.,  65.,  27.,   7.,  16.,   2.,  28.,
         6., 123., 120.,   7.,   3.,   0.,  81.,   2.,  54., 122.,   4.,
        14.,   0., 100.,  15.,   0.,  57.,  99.,  37.,  38.,  32.,  21.,
        32.,  40.,   0.,   5.,   8.,   5.,   0.,  50.,  30.,  37.,  89.,
         4.,  98.,  83.,  93.,   0.,  52., 152.,   1.,   8.,  93.,  45.,
        26.,   0.,   1.,   0.,  16.,  47.,  89.,   3.,   1.,  53.,  16.,
         0.,  81.,  14.,  78.,   6., 105., 122.,   9.,   8.,  28.,  35.,
        69.,  13.,  97.,  93.,   2.,  36.,  39.,   2.,  29.,  12.,  19.,
        34.,   2., 100.,  44.,  82.,   0.,  79.,   6.,   9.,   8.,  23.,
        93.,  35.,  63.,  74.,   8., 117.,  39.,  4

In [54]:
sachin_cumscores = np.cumsum(sachin)

In [55]:
np.histogram(sachin_cumscores, bins=np.arange(0, 10000, 1000))

(array([29, 18, 26, 25, 26, 26, 23, 22, 30], dtype=int64),
 array([   0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000]))